src/trainers/trainer.cpp

```diff
@@ -144,26 +144,27 @@ trainer::execution_context_key_pair_t trainer::check_and_build_execution_context
     if(dynamic_cast<observer_ptr<sgd_training_algorithm>>(&alg) != nullptr) {
       /// @todo BVE FIXME Figure out how to get a good mini-batch size
       /// in here
-      context = make_unique<sgd_execution_context>(this, m_comm, mode, model->get_max_mini_batch_size());
+      context = make_unique<sgd_execution_context>(this, &alg, m_comm, mode, model->get_max_mini_batch_size());
     }else {
-      context = make_unique<execution_context>(this, m_comm, mode);
+      context = make_unique<execution_context>(this, &alg, m_comm, mode);
     }
     m_model_execution_context.emplace(key,std::move(context));
   }
   return key;
 }
 
 /// Check if there is already an execution context for the model in this mode, if not create one
-trainer::execution_context_key_pair_t trainer::check_and_build_execution_context(const execution_context& c,
+trainer::execution_context_key_pair_t trainer::check_and_build_execution_context(execution_context& c,
                                                                                  model& model,
                                                                                  execution_mode mode) {
   auto key = std::make_pair(&model, mode);
   if(m_model_execution_context.count(key) == 0) {
     std::unique_ptr<execution_context> context;
-    if(dynamic_cast<observer_ptr<const sgd_execution_context>>(&c) != nullptr) {
-      context = make_unique<sgd_execution_context>(this, m_comm, mode, model.get_max_mini_batch_size());
+    //    observer_ptr<training_algorithm> alg = const_cast
+    if(dynamic_cast<observer_ptr</*const */sgd_execution_context>>(&c) != nullptr) {
+      context = make_unique<sgd_execution_context>(this, &(c.get_training_algorithm()), m_comm, mode, model.get_max_mini_batch_size());
```

Comment:
This shouldn't work with pointers -- prefer references unless you can meaningfully pass `nullptr`. The address-of operator here is clunky at best.